### ISSUE TITLE
Function Cast is not accessible (in codegen)

### ISSUE THREAD

**Issue by BenjaminGibbs** [dotnet/orleans#71]
I am getting the error below when I compile one of my iGrain interfaces.

```
Error	6	'Orleans.GrainFactory.Friend Shared Function Cast(Of TGrainInterface)(grain As Orleans.Runtime.IAddressable) As TGrainInterface' is not accessible in this context because it is 'Friend'.	C:\Users\Benjamin\Documents\Visual Studio 2013\Projects\KeynotelyOrleans\Keynotely.GrainInterfaces\GeneratedFiles\orleans.codegen.vb	1689	190	Keynotely.GrainInterfaces
```

What appears to be the trigger of this error is if I pass an iGrain in to a parameter of a function in the iGrain interface.

For example:

```
Function SaveIndexAsync(grainType As String, propertyName As String, propertyValue As Object, grainReference As IGrain) As task
```

The GrainFactory isn't exposing the Cast function, where as GrainFactoryWith*****Key all expose the cast, for example if I updated my code to:

```
Function SaveIndexAsync(grainType As String, propertyName As String, propertyValue As Object, grainReference As IGrainWithStringKey) As task
```

It will compile fine, I am not sure if this a bug, or is by design - any input or thoughts are welcome.

**Comment by sergeybykov** [comment]
Background. This line of generated code specifically protects from passing a *this* reference as an argument to a grain method call. We've seen people doing it by mistake early on enough to add this protection. To pass a reference to itself, a grain needs to pass this.AsReference(), which is an un-typed instance of GrainReference class. Hence it needs the cast.

Declaring the argument type as IGrain creates an interesting challenge because GrainReference doesn't implement IGrain. It implements IAddressable, which is a parent of IGrain. If I change the argument type from IGrain to IAddressable, it compiles fine, but this protection line doesn't get generated, which isn't right. Need to dig more.

**Comment by gabikliot** [comment]
Fixed via #288.
